libavcodec/mfdec.c

```diff
@@ -0,0 +1,1717 @@
+        return AVERROR(ENOMEM);
+    }
+
+    if ((ret = mf_unlock_async(avctx)) < 0)
+        return ret;
+
+    hr = IMFTransform_QueryInterface(c->mft, &IID_ICodecAPI, (void **)&c->codec_api);
+    if (!FAILED(hr))
+        av_log(avctx, AV_LOG_VERBOSE, "MFT supports ICodecAPI.\n");
+
+    if (c->is_dec) {
+        const char *bsf = NULL;
+
+        if (avctx->codec->id == AV_CODEC_ID_H264 && avctx->extradata && avctx->extradata[0] == 1)
+            bsf = "h264_mp4toannexb";
+
+        if (avctx->codec->id == AV_CODEC_ID_HEVC && avctx->extradata && avctx->extradata[0] == 1)
+            bsf = "hevc_mp4toannexb";
+
+        if (bsf) {
+            const AVBitStreamFilter *bsfc = av_bsf_get_by_name(bsf);
+            if (!bsfc) {
+                ret = AVERROR(ENOSYS);
+                goto bsf_done;
+            }
+            if ((ret = av_bsf_alloc(bsfc, &c->bsfc)) < 0)
+                goto bsf_done;
+            if ((ret = avcodec_parameters_from_context(c->bsfc->par_in, avctx)) < 0)
+                goto bsf_done;
+            if ((ret = av_bsf_init(c->bsfc)) < 0)
+                goto bsf_done;
+            ret = 0;
+        bsf_done:
+            if (ret < 0) {
+                av_log(avctx, AV_LOG_ERROR, "Cannot open the %s BSF!\n", bsf);
+                return ret;
+            }
+        }
+    }
+
+    if (c->is_video && ((ret = mf_init_hwaccel(avctx)) < 0))
+        return ret;
+
+    if ((ret = mf_check_codec_requirements(avctx)) < 0)
+        return ret;
+
+    hr = IMFTransform_GetStreamIDs(c->mft, 1, &c->in_stream_id, 1, &c->out_stream_id);
+    if (hr == E_NOTIMPL) {
+        c->in_stream_id = c->out_stream_id = 0;
+    } else if (FAILED(hr)) {
+        av_log(avctx, AV_LOG_ERROR, "could not get stream IDs (%s)\n", ff_hr_str(hr));
+        return AVERROR_EXTERNAL;
+    }
+
+    if ((ret = mf_negotiate_types(avctx)) < 0)
+        return ret;
+
+    if ((ret = mf_setup_context(avctx)) < 0)
+        return ret;
+
+    hr = IMFTransform_ProcessMessage(c->mft, MFT_MESSAGE_NOTIFY_BEGIN_STREAMING, 0);
+    if (FAILED(hr)) {
+        av_log(avctx, AV_LOG_ERROR, "could not start streaming (%s)\n", ff_hr_str(hr));
+        return AVERROR_EXTERNAL;
+    }
+
+    hr = IMFTransform_ProcessMessage(c->mft, MFT_MESSAGE_NOTIFY_START_OF_STREAM, 0);
+    if (FAILED(hr)) {
+        av_log(avctx, AV_LOG_ERROR, "could not start stream (%s)\n", ff_hr_str(hr));
+        return AVERROR_EXTERNAL;
+    }
+
+    c->lavc_init_done = 1;
+
+    return 0;
+}
+
+static int mf_close(AVCodecContext *avctx)
+{
+    MFContext *c = avctx->priv_data;
+    int uninit_com = c->mft != NULL;
+
+    if (c->codec_api)
+        ICodecAPI_Release(c->codec_api);
+
+    if (c->async_events)
+        IMFMediaEventGenerator_Release(c->async_events);
+
+    av_bsf_free(&c->bsfc);
+
+    av_buffer_unref(&c->frames_ref);
+    av_frame_free(&c->tmp_frame);
+    av_buffer_unref(&c->decoder_ref);
+
+    if (uninit_com)
+        CoUninitialize();
+
+    if (c->is_enc) {
```

Comment:
Do we need these `c->is_enc` branches in here? Will that field ever be true? The mfenc.c file doesn't seem to check `c->is_enc / c->is_dec` at all

Reply:
Good catch, the is_enc/is_dec was a holdover from the original Pawel Wegner patch source, when encoding and decoding used to be included in the same source.  I have cleaned this up in the next commit.